### PR TITLE
chore(packaging): Migrate from `setuptools` to `hatchling`

### PR DESCRIPTION
**Pull request by webknjaz**
This is another modern build backend. The patch deletes `setup.py` from the repository, and adds configuration for `hatchling` and declares PEP 621 metadata in `pyproject.toml`.

It does not attempt to change the previously existing behavior.

Put an `x` into the box if that apply:

- [ ] This PR introduces breaking change.
- [ ] This PR fixes a bug.
- [ ] This PR adds new functionality.
- [x] This PR enhances existing functionality.

### Description of your changes

$sbj.

### How can we test changes

`python -Im build`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  'Programming Language :: Python :: Implementation :: PyPy',
+]
+description = 'Pre-commit hooks for terraform_docs_replace'
+dependencies = [
+  'setuptools-git-version',
+]
+
+[[project.authors]]
+name = 'Contributors'  # FIXME
+# email = 'степан@криївка.укр'
```
Comment (MaxymVlasov) [inline pyproject.toml:28]: Ну не Гіга і на тому спасибі :D

**Review by MaxymVlasov — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (webknjaz) [reply pyproject.toml:28]: Та тут просто взагалі трохи дивне значення в оригіналі. Потім треба нормально заповнити authors/maintainers. Але я не хотів цим займатися зараз, лише дуже мінімальна міграція.

**Comment by yermulnik** [comment]
@webknjaz @MaxymVlasov Would it make sense 1) to add a brief outline to this PR description of what is changing for end users (if any), what benefits this change adds to `pre-commit-terraform` overall and 2) to add a brief about to README (if applicable)? 🤔 
I mean that «_This is another modern build backend_» is kinda pretty much concise and doesn't explain why.
Thanks.

**Comment by MaxymVlasov** [comment]
1. No changes to end users
2. No xhanges in Readme required


This actually affects only how hooks written on python discovered by other python tools (in our case pre-commit itself)

Currently there only 1 python hook - terraform_docs_replace, deprecated for years. 
So, TL;DR: it shoudn't affect anything, but if so, only deprecated and marked for removal long time ago hook could be broken

**Comment by webknjaz** [comment]
We talked about it privately so some context might be missing. But essentially nothing changes for anyone really.
It's an internal implementation detail and only changes what `pip install .` calls under the hood.
Just treat it as making dynamic packaging declaration static through declarative configs.

File: pyproject.toml (hunk 1 above)
Comment (yermulnik) [reply pyproject.toml:28]: Let's stick with ASCII please since it's not all of the `pre-commit-terraform` users who can read non-English.

Comment (webknjaz) [reply pyproject.toml:28]: It's just a code comment. Someone should change it to real values.

**Comment by yermulnik** [comment]
@webknjaz @MaxymVlasov Thanks for details and context. Let's anyways add details in PRs description so that other contributors have an idea on the context of private discussions 🤝

**Comment by MaxymVlasov** [comment]
About benefits @webknjaz can provide better info, but AFAIK, setup.py i this repo actually configured in wrong way from day 1

And this PR with bunch of others, make python hooks more isolated and predictable during direct calls to hooks via test execution, for instance.

And just get rid of of legacy stuff which just don't work

File: pyproject.toml (hunk 1 above)
Comment (yermulnik) [reply pyproject.toml:28]: It's just a code comment that is unreadable by English-only users. Most of the time it's just a garbage symbols for them, you know.

Comment (yermulnik) [reply pyproject.toml:28]: > Someone should change it to real values.

Until then, it's best to replace with ASCII chars or remove. It's just an opinion of mine though. I'll deffer this to @MaxymVlasov as main collaborator.

**Comment by yermulnik** [comment]
> About benefits @webknjaz can provide better info, but AFAIK, setup.py in this repo actually configured in wrong way from day 1
> 
> And this PR with bunch of others, make python hooks more isolated and predictable during direct calls to hooks via test execution, for instance.
> 
> And just get rid of of legacy stuff which just don't work

Yep-yep, already got the idea =) My point is to share context with PR description right away =)

**Comment by webknjaz** [comment]
@yermulnik I'll try adding more context in general. It's just that Max asked for help, which lead to me discovering a number of typical structural and infra problems in the repo. So I wanted to make a few small PRs to address that but I don't have a lot of time to dedicate to this.

**Comment by yermulnik** [comment]
> @yermulnik I'll try adding more context in general.

Appreciate it 🤝 

> It's just that Max asked for help, […] but I don't have a lot of time to dedicate to this.

Didn't want to add burden for you. Just am carrying about future reference and posterity.
Where you see a concise description as a bit of overhead, I'm all fine with something like «_Discussed with @MaxymVlasov , blame that guy if anything_» 😂

**Comment by antonbabenko** [comment]
This PR is included in version 1.97.0 :tada:

**Pull request merged**